### DISCUSSION TITLE
Flash attention

### DISCUSSION THREAD

**Discussion by pfeatherstone** [tinygrad/tinygrad#1505] [General]
Is something like Flash Attention (2) automatically computed by tinygrad with all the lazy (expression template ?) stuff?

**Comment by terafo** [comment]
No, Flash Attention isn't inferred by Tinygrad yet.

**Comment by chenyuxyz** [comment]
attention without mask and dropout (just `softmax(q@k)@v`) currently has 6 kernels: 1 matmul, 4 for softmax, 1 matmul. To squash softmax into one, we need a generalized unary op fusion like norm fusion. Then it's equivalent to fusing 2 matmuls (reduces) into one.
